### ISSUE TITLE
Weird behaviour when changing predicate of MGLCircleStyleLayer at runtime

### ISSUE THREAD

**Issue by rmnblm** [mapbox/mapbox-gl-native#6478]
**Platform:** iOS
**Mapbox SDK version:** v3.4.0-alpha.4

I'm playing around with runtime styling for a while now and I'm facing some weird problems when I change the predicate of an MGLCircleStyleLayer at runtime. To get a better idea of what I mean, you can clone [this repo](https://github.com/rmnblm/mapbox-ios-examples/tree/05e9ccaa197b885e45308398b8d9df054b70d185) and try "Semantic Zooming" from the example list.
### Steps to trigger behavior
1. I want to show more or less details based on the `mapView.zoomLevel`. I implemented this like [this](https://github.com/rmnblm/mapbox-ios-examples/blob/05e9ccaa197b885e45308398b8d9df054b70d185/src/Examples/SemanticZoomingViewController.swift) and [this](https://github.com/rmnblm/mapbox-ios-examples/blob/05e9ccaa197b885e45308398b8d9df054b70d185/src/Examples/SemanticZoomingDataSource.swift)
2. Every time the mapView did change, I'm revalidating my layers if they should be visible or not (based on the zoomLevel)

```
func validate(zoomLevel: Double) {
    if zoomLevel >= visibleAt {
        styleLayer?.predicate = predicate
    } else {
        // Predicate always returns false to hide layer
        // Sadly, NSPredicate(value: false) doesn't work with Mapbox
        styleLayer?.predicate = NSPredicate(format: "test == 'not'")
    }
}
```
1. If I'm running the app, layers really are turned off and on when I zoom in, but weird behaviour occurs when I zoom out. For a better understanding, please watch the following gif.

![out](https://cloud.githubusercontent.com/assets/5942764/18873351/46e5778a-84bf-11e6-9459-de1c588a3577.gif)

The green dots appear when zooming in, disappear when zooming out a little bit and re-appear again when zooming out a little bit more.

**Comment by rmnblm** [comment]
I tried to implement it in a different way by adding/removing the layer completely and not just changing the predicate at runtime. I still have the same behaviour, see implementation details [here](https://github.com/rmnblm/mapbox-ios-examples/blob/ac457f870aaf571346e97a0b8fe10684dbf4d4bf/src/Examples/SemanticZoomingDataSource.swift) and [here](https://github.com/rmnblm/mapbox-ios-examples/blob/ac457f870aaf571346e97a0b8fe10684dbf4d4bf/src/Examples/SemanticZoomingViewController.swift). 

Has this something to do with https://github.com/mapbox/mapbox-gl-native/issues/6361 because the mapView is loading the styles asynchronously?

**Comment by jfirebaugh** [comment]
There may be a bug here, but the recommended way to accomplish this use case is to set the `minimumZoomLevel` property. That informs Mapbox GL of your intent, and it can then optimize accordingly.

**Comment by rmnblm** [comment]
I fixed this problem by adding/removing the layers in the `func mapViewDidFinishRenderingFrame(_ mapView: MGLMapView, fullyRendered: Bool)` method.
Does this help you finding the bug?

You can see full implementation details [here](https://github.com/rmnblm/mapbox-ios-examples/blob/master/src/Examples/SemanticZoomingViewController.swift)

**Comment by 1ec5** [comment]
> Sadly, NSPredicate(value: false) doesn't work with Mapbox

This was fixed in #6405, which hasn’t made it into a prerelease yet.
